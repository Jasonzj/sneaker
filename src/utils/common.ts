/**
 * 格式化时间戳
 *
 * @param {number} time 时间戳
 * @returns {string}
 */
export const formatTime = (time: number): string => {
  const date = new Date(time + 8 * 3600 * 1000)
  return date.toJSON().substr(0, 19).replace('T', ' ')
}

/**
 * 是否为空对象
 *
 * @param {(Record<string, unknown> | undefined)} o
 * @returns {boolean}
 */
export const isEmptyObject = (o: Record<string, unknown> | undefined | null): boolean => {
  if (!o) return true
  return Object.keys(o).length === 0
}

export const searchJumpScroll = (): void => window.scrollTo({ top: 580, behavior: 'smooth' })
