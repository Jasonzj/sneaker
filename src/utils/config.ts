const interfaceName = {
  stockx: 'stockx',
  goat: 'goat',
  dewu: 'dewu',
  db: 'db',
}

const defaultInterface = interfaceName.dewu

export default {
  title: 'Sneakers App',
  bannerText: 'Search for Shoes and Integration Price For All Size',
  openPages: ['/sign_in', '/sign_up'],
  homePage: '/trending/dewu',
  apiPrefix: '/api/v1',
  interfaceName,
  defaultInterface,
}
