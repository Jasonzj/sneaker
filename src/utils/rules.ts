export const usernameRules = {
  required: 'This is required.',
  minLength: {
    value: 5,
    message: 'The minimum length cannot exceed 5',
  },
  maxLength: {
    value: 20,
    message: 'The maximum length cannot exceed 20',
  },
}

export const passwordRules = {
  ...usernameRules,
  pattern: {
    value: /^(?![0-9]+$)(?![a-zA-Z]+$)[0-9A-Za-z]{1,}$/,
    message: 'At least one number and letter and no special characters',
  },
}
