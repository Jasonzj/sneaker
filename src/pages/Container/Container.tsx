import React from 'react'
import { Switch, Route, Redirect, useRouteMatch } from 'react-router-dom'
import { CSSTransition, SwitchTransition } from 'react-transition-group'

// components
import Product from '../Product'
import NoMatch from '../NoMatch/'
import Header from '../../components/Header'
import Banner from '../../components/Banner'
import SignIn from '../SignIn'
import SignUp from '../SignUp'

// config
import config from '../../utils/config'
const { homePage, openPages } = config

const openPagesMap = {
  '/sign_in': <Route exact path='/sign_in' component={SignIn} />,
  '/sign_up': <Route exact path='/sign_up' component={SignUp} />,
}

const Container: React.FC = () => {
  const match = useRouteMatch({
    path: ['/sign_in', '/sign_up'],
    exact: true,
  })

  if (match && openPages && openPages.includes(match.path)) {
    return (
      <SwitchTransition mode='out-in'>
        <CSSTransition key={match.path} timeout={1000} in classNames='sign-ani'>
          {openPagesMap[match.path as '/sign_in' | '/sign_up']}
        </CSSTransition>
      </SwitchTransition>
    )
  }

  return (
    <div className='App'>
      <Header />
      <Banner />
      <Switch>
        <Route exact path='/' render={() => <Redirect to={homePage} />} />
        <Route exact path='/trending/:siteName' component={Product} />
        <Route exact path='/search/:siteName/:key' component={Product} />
        <Route exact path='/following' component={Product} />
        <Route component={NoMatch} />
      </Switch>
    </div>
  )
}

export default Container
