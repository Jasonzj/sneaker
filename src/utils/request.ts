import axios, { AxiosError } from 'axios'
import { createHashHistory } from 'history'
import message from '../components/Message'
import config from './config'

const history = createHashHistory()

const request = axios.create({
  baseURL: `${config.apiPrefix}`,
  timeout: 1000 * 30,
})

request.interceptors.request.use(
  function (config) {
    const token = window.localStorage.getItem('token')
    if (token) config.headers.Authorization = `Bearer ${window.localStorage.getItem('token')}`
    return config
  },
  function (error: AxiosError) {
    return Promise.reject(error)
  },
)

request.interceptors.response.use(
  function (response) {
    return response
  },
  function (error: AxiosError) {
    if (error.response?.status === 401) {
      const token = window.localStorage.getItem('token')
      token && message.error('Login information is invalid, please login again!')
      window.localStorage.removeItem('token')
      history.push('/sign_in')
    }

    if (error?.code === 'ECONNABORTED' && error.message.includes('timeout')) {
      message.error('Request Timeout!')
    }

    return Promise.reject(error)
  },
)

export default request
