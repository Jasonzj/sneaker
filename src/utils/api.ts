import { Method } from 'axios'
import request from './request'
import { ApiLoaderType } from './global'

enum methods {
  get = 'get',
  post = 'post',
  put = 'put',
  del = 'delete',
}

const setApiLoader = (api: string, method: Method): ApiLoaderType => {
  return async (params, cancelToken) => {
    const getParams = params && method === methods.get ? `/${Object.values(params).join('/')}` : ''
    return await request({
      method,
      url: `${api}${getParams}`,
      data: getParams ? null : params,
      cancelToken,
    })
  }
}

const getProductPrice = `/prices`
const getProductLists = `/search`
const getTrendingLists = `/trending`
const getFollowingLists = `/user/following`
const getFollowingDetailLists = `/user/following?isDetail=true`
const getSearchSuggestions = `/suggestions`
const login = `/login`
const register = `/register`

export const ProductListsLoader = setApiLoader(getProductLists, methods.get)
export const ProductPriceLoader = setApiLoader(getProductPrice, methods.get)
export const TrendingListsLoader = setApiLoader(getTrendingLists, methods.get)
export const FollowLoader = setApiLoader(getFollowingLists, methods.put)
export const unFollowLoader = setApiLoader(getFollowingLists, methods.del)
export const FollowingListsLoader = setApiLoader(getFollowingLists, methods.get)
export const FollowingDetailListsLoader = setApiLoader(getFollowingDetailLists, methods.get)
export const SearchSuggestionsLoader = setApiLoader(getSearchSuggestions, methods.get)
export const LoginLoader = setApiLoader(login, methods.post)
export const RegisterLoader = setApiLoader(register, methods.post)
