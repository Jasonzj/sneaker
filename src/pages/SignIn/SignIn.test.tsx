import user from '@testing-library/user-event'
import { render, screen, waitForElementToBeRemoved } from '../../tests/routerRender'
import Container from '../Container/'
import { createHashHistory } from 'history'
import { usernameRules, passwordRules } from '../../utils/rules'
import { login_default, login_error } from '../../tests/test.mock'
import { rest, server } from '../../tests/testSever'

beforeEach(() => {
  jest.useFakeTimers()
})

const history = createHashHistory()
history.push('/sign_in')

test('should render correctly rules of the error', async () => {
  const { asFragment } = render(<Container />)
  const Username_Input = screen.getByPlaceholderText('username')
  const Password_Input = screen.getByPlaceholderText('password')

  user.type(Username_Input, 'a')
  user.type(Password_Input, 'a')

  expect(await screen.findAllByText(usernameRules.minLength.message)).toHaveLength(2)
  expect(Username_Input).toHaveClass('error')
  expect(Password_Input).toHaveClass('error')
  expect(asFragment()).toMatchSnapshot()

  user.type(Username_Input, 'dmin')
  user.type(Password_Input, 'dmin')

  expect(await screen.findByText(passwordRules.pattern.message)).toBeInTheDocument()
  expect(screen.queryByText(usernameRules.minLength.message)).not.toBeInTheDocument()
  expect(Username_Input).not.toHaveClass('error')
  expect(Password_Input).toHaveClass('error')
  expect(asFragment()).toMatchSnapshot()

  user.type(Username_Input, '0123456789012345')
  user.type(Password_Input, '0123456789012345')

  expect(await screen.findAllByText(usernameRules.maxLength.message)).toHaveLength(2)
  expect(Username_Input).toHaveClass('error')
  expect(Password_Input).toHaveClass('error')
  expect(asFragment()).toMatchSnapshot()

  user.clear(Username_Input)
  user.clear(Password_Input)

  expect(await screen.findAllByText(usernameRules.required)).toHaveLength(2)
  expect(Username_Input).toHaveClass('error')
  expect(Password_Input).toHaveClass('error')
  expect(asFragment()).toMatchSnapshot()

  user.type(Username_Input, 'admin')
  user.type(Password_Input, 'admin1')

  await waitForElementToBeRemoved(() => screen.getAllByText(usernameRules.required))
  expect(screen.queryByText(usernameRules.minLength.message)).not.toBeInTheDocument()
  expect(screen.queryByText(usernameRules.maxLength.message)).not.toBeInTheDocument()
  expect(screen.queryByText(passwordRules.pattern.message)).not.toBeInTheDocument()
  expect(Username_Input).not.toHaveClass('error')
  expect(Password_Input).not.toHaveClass('error')
  expect(asFragment()).toMatchSnapshot()
})

test('should render correctly class with page changed', async () => {
  const { asFragment } = render(<Container />)

  user.click(screen.getByText('Sign Up'))

  expect(await screen.findByText('Sign Up')).toHaveClass('active')
  expect(asFragment()).toMatchSnapshot()
})

test('should handle correctly login', async () => {
  history.push('/sign_in')
  const { asFragment, rerender } = render(<Container />)

  user.type(screen.getByPlaceholderText('username'), 'admin')
  user.type(screen.getByPlaceholderText('password'), 'admin1')
  user.click(screen.getByRole('button', { name: 'Login In' }))

  expect(await screen.findByTestId('product_card')).toBeInTheDocument()
  expect(window.localStorage.getItem('token')).toBe(login_default.data.token)
  expect(window.localStorage.getItem('name')).toBe(login_default.data.username)
  expect(asFragment()).toMatchSnapshot()

  server.use(
    rest.get('/api/v1/trending/dewu', (req, res, ctx) => {
      return res(ctx.status(401), ctx.json(login_error))
    }),
  )

  history.push('/sign_in')
  rerender(<Container />)

  user.type(await screen.findByPlaceholderText('username'), 'admin')
  user.type(screen.getByPlaceholderText('password'), 'admin1')
  user.click(screen.getByRole('button', { name: 'Login In' }))

  expect(await screen.findByText(login_error.msg)).toBeInTheDocument()
  expect(asFragment()).toMatchSnapshot()
})
