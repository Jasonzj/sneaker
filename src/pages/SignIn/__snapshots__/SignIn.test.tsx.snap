// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle correctly login 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <header
      class="header"
    >
      <div
        class="logoBox"
      >
        <h1>
          Sneakers App
        </h1>
        <img
          alt="logo"
          class="logo"
          src="logo.png"
        />
      </div>
      <nav
        class="nav"
      >
        <a
          href="#/trending/dewu"
        >
          Dewu Trending
        </a>
        <a
          href="#/trending/goat"
        >
          Goat Trending
        </a>
        <div
          class="main"
        >
          <div
            class="content"
          >
            <div
              class="avatar"
            >
              <img
                src="https://api.prodless.com/avatar.png"
              />
            </div>
          </div>
        </div>
      </nav>
    </header>
    <div
      class="banner"
    >
      <div
        class="introduce"
      >
        Search for Shoes and Integration Price For All Size
      </div>
      <div
        class="searchBox"
      >
        <input
          class="searchInput"
          title="search_input"
          type="text"
          value=""
        />
        <button
          class="searchButton"
          title="search_button"
        />
        <div
          class="selectBox"
        >
          <span>
            Search Interface: 
          </span>
          <select
            class="searchInterface"
            title="search_interface_select"
          >
            <option
              value="dewu"
            >
              Dewu
            </option>
            <option
              value="stockx"
            >
              StockX
            </option>
            <option
              value="goat"
            >
              Goat
            </option>
            <option
              value="db"
            >
              DB
            </option>
          </select>
          <span>
            Search Suggestions Interface: 
          </span>
          <select
            class="searchInterface"
          >
            <option
              value="dewu"
            >
              Dewu
            </option>
            <option
              value="stockx"
            >
              StockX
            </option>
          </select>
        </div>
        <div
          class="searchSuggestion"
        />
      </div>
      <div
        class="iconBar"
      >
        <img
          alt="goat"
          class="logo"
          src="goat.png"
        />
        <img
          alt="dewu"
          class="logo"
          src="dewu.png"
        />
        <img
          alt="stockx"
          class="logo"
          src="stockx.png"
        />
      </div>
    </div>
    <div
      class="container"
    >
      <div
        class="header"
      >
        <h1
          class="title"
          title="search_title"
        >
          Trending(dewu)
        </h1>
      </div>
      <div
        class="box"
      >
        <div
          class="loader hidden"
          title="loader"
        >
          <div
            class="wrapper"
          >
            <div
              class="inner"
            />
            <div
              class="text"
            >
              LOADING
            </div>
          </div>
        </div>
        <div
          class="card"
          data-testid="product_card"
        >
          <img
            alt="Air Jordan 1 Mid Shattered Backboard 扣碎篮板 篮球鞋"
            src="https://cdn.poizon.com/temp/Fn20HZPjmCnQk_VCNJ9C2HRg0xP5"
          />
          <div
            class="shoeName"
          >
            Air Jordan 1 Mid Shattered Backboard 扣碎篮板 篮球鞋
          </div>
          <div
            class="price"
          >
            $109
          </div>
          <div
            class="brand"
          >
            stockx
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`should handle correctly login 2`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="main"
    >
      <div
        class="title"
      >
        <a
          class="active"
          href="#/sign_in"
        >
          Sign In
        </a>
        <b>
          ·
        </b>
        <a
          href="#/sign_up"
        >
          Sign Up
        </a>
      </div>
      <form
        class="form"
      >
        <div
          class="inputBox"
        >
          <input
            class="input"
            name="username"
            placeholder="username"
            type="text"
          />
        </div>
        <div
          class="inputBox"
        >
          <input
            class="input"
            name="password"
            placeholder="password"
            type="password"
          />
        </div>
        <button
          class="button"
          type="submit"
        >
          <div
            class="content"
          >
            Login In
          </div>
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`should render correctly class with page changed 1`] = `
<DocumentFragment>
  <div
    class="container sign-ani-enter sign-ani-enter-active"
  >
    <div
      class="main"
    >
      <div
        class="title"
      >
        <a
          href="#/sign_in"
        >
          Sign In
        </a>
        <b>
          ·
        </b>
        <a
          class="active"
          href="#/sign_up"
        >
          Sign Up
        </a>
      </div>
      <form
        class="form"
      >
        <div
          class="inputBox"
        >
          <input
            class="input"
            name="username"
            placeholder="username"
            type="text"
          />
        </div>
        <div
          class="inputBox"
        >
          <input
            class="input"
            name="password"
            placeholder="password"
            type="password"
          />
        </div>
        <button
          class="button"
          type="submit"
        >
          <div
            class="content"
          >
            Register
          </div>
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`should render correctly rules of the error 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="main"
    >
      <div
        class="title"
      >
        <a
          class="active"
          href="#/sign_in"
        >
          Sign In
        </a>
        <b>
          ·
        </b>
        <a
          href="#/sign_up"
        >
          Sign Up
        </a>
      </div>
      <form
        class="form"
      >
        <div
          class="inputBox"
        >
          <input
            class="input error"
            name="username"
            placeholder="username"
            type="text"
          />
          <p
            class="errorMsg"
          >
            The minimum length cannot exceed 5
          </p>
        </div>
        <div
          class="inputBox"
        >
          <input
            class="input error"
            name="password"
            placeholder="password"
            type="password"
          />
          <p
            class="errorMsg"
          >
            The minimum length cannot exceed 5
          </p>
        </div>
        <button
          class="button"
          type="submit"
        >
          <div
            class="content"
          >
            Login In
          </div>
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`should render correctly rules of the error 2`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="main"
    >
      <div
        class="title"
      >
        <a
          class="active"
          href="#/sign_in"
        >
          Sign In
        </a>
        <b>
          ·
        </b>
        <a
          href="#/sign_up"
        >
          Sign Up
        </a>
      </div>
      <form
        class="form"
      >
        <div
          class="inputBox"
        >
          <input
            class="input"
            name="username"
            placeholder="username"
            type="text"
          />
        </div>
        <div
          class="inputBox"
        >
          <input
            class="input error"
            name="password"
            placeholder="password"
            type="password"
          />
          <p
            class="errorMsg"
          >
            At least one number and letter and no special characters
          </p>
        </div>
        <button
          class="button"
          type="submit"
        >
          <div
            class="content"
          >
            Login In
          </div>
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`should render correctly rules of the error 3`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="main"
    >
      <div
        class="title"
      >
        <a
          class="active"
          href="#/sign_in"
        >
          Sign In
        </a>
        <b>
          ·
        </b>
        <a
          href="#/sign_up"
        >
          Sign Up
        </a>
      </div>
      <form
        class="form"
      >
        <div
          class="inputBox"
        >
          <input
            class="input error"
            name="username"
            placeholder="username"
            type="text"
          />
          <p
            class="errorMsg"
          >
            The maximum length cannot exceed 20
          </p>
        </div>
        <div
          class="inputBox"
        >
          <input
            class="input error"
            name="password"
            placeholder="password"
            type="password"
          />
          <p
            class="errorMsg"
          >
            The maximum length cannot exceed 20
          </p>
        </div>
        <button
          class="button"
          type="submit"
        >
          <div
            class="content"
          >
            Login In
          </div>
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`should render correctly rules of the error 4`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="main"
    >
      <div
        class="title"
      >
        <a
          class="active"
          href="#/sign_in"
        >
          Sign In
        </a>
        <b>
          ·
        </b>
        <a
          href="#/sign_up"
        >
          Sign Up
        </a>
      </div>
      <form
        class="form"
      >
        <div
          class="inputBox"
        >
          <input
            class="input error"
            name="username"
            placeholder="username"
            type="text"
          />
          <p
            class="errorMsg"
          >
            This is required.
          </p>
        </div>
        <div
          class="inputBox"
        >
          <input
            class="input error"
            name="password"
            placeholder="password"
            type="password"
          />
          <p
            class="errorMsg"
          >
            This is required.
          </p>
        </div>
        <button
          class="button"
          type="submit"
        >
          <div
            class="content"
          >
            Login In
          </div>
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`should render correctly rules of the error 5`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="main"
    >
      <div
        class="title"
      >
        <a
          class="active"
          href="#/sign_in"
        >
          Sign In
        </a>
        <b>
          ·
        </b>
        <a
          href="#/sign_up"
        >
          Sign Up
        </a>
      </div>
      <form
        class="form"
      >
        <div
          class="inputBox"
        >
          <input
            class="input"
            name="username"
            placeholder="username"
            type="text"
          />
        </div>
        <div
          class="inputBox"
        >
          <input
            class="input"
            name="password"
            placeholder="password"
            type="password"
          />
        </div>
        <button
          class="button"
          type="submit"
        >
          <div
            class="content"
          >
            Login In
          </div>
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
